internal/tests/serialization/set.go

```diff
@@ -0,0 +1,223 @@
+package serialization
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"reflect"
+	"runtime/debug"
+	"testing"
+)
+
+// Set is a tool for marshal and unmarshall funcs testing.
+type Set struct {
+	Data   []byte
+	Values []interface{}
+
+	BrokenMarshalTypes   []reflect.Type
+	BrokenUnmarshalTypes []reflect.Type
+}
+
+// RunTest runs tests for cases when the function should no error,
+// on marshal - marshaled data from Set.Values should be equal with Set.Data,
+// on unmarshall - unmarshalled value from Set.Data should be equal with Set.Values.
+func (s Set) RunTest(name string, t *testing.T, marshal func(interface{}) ([]byte, error), unmarshal func([]byte, interface{}) error) {
+	if name == "" {
+		t.Fatal("name should be provided")
+	}
+
+	t.Run(name, func(t *testing.T) {
+		for i := range s.Values {
+			val := s.Values[i]
+
+			t.Run(fmt.Sprintf("%T", val), func(t *testing.T) {
+				if marshal != nil {
+					s.runMarshalTest(t, marshal, val)
+				}
+
+				if unmarshal != nil {
+					if rt := reflect.TypeOf(val); rt.Kind() != reflect.Ptr {
+						unmarshalIn := newRef(val)
+						s.runUnmarshalTest("unmarshal", t, unmarshal, val, unmarshalIn)
+					} else {
+						// Test unmarshal to (*type)(nil)
+						unmarshalIn := newRef(val)
+						s.runUnmarshalTest("unmarshal**nil", t, unmarshal, val, unmarshalIn)
+
+						// Test unmarshal to &type{}
+						unmarshalInZero := newRefToZero(val)
+						s.runUnmarshalTest("unmarshal**zero", t, unmarshal, val, unmarshalInZero)
+					}
+				}
+			})
+		}
+	})
+}
+
+// RunCorruptTest runs tests for cases when the function should an error.
+func (s Set) RunCorruptTest(name string, t *testing.T, marshal func(interface{}) ([]byte, error), unmarshal func([]byte, interface{}) error) {
```

Comment:
In order to avoid confusiton let's have another structure for these tests or, even split it into two (for negative serialization and deserialization tests), since you already logically split them:
`Set` -> `PositiveSerializationSet`
`NegativeMarshalSet`
`NetativeUnmarhsalSet`

Feel free to suggest better names for them, or provide contr argument why we hsould not do it.

Reply:
:) It's more difficult than just writing code.

Reply:
Done